### PR TITLE
CodeGen: Avoid double cleanup when copying to memory

### LINKED ISSUE #8670: Optimiser confused by something
This is the most minimal example I've got so far:
```
contract C {
  function f() external {
        bytes memory amount;
        bytes24 zero_bytes24;
        bytes32 signature_root;
        bytes32 node = sha256(abi.encodePacked(
            sha256(abi.encodePacked(amount, zero_bytes24, signature_root))
        ));
  }
}
```

With `--optimize` (default settings, albeit tried with `--optimize-runs 10000000` too) there is this snippet in the output:
```
        /* "optim.sol":197:251  abi.encodePacked(amount, zero_bytes24, signature_root) */
      pop
      pop
      pop
      swap1
      pop
      add
      dup4
      not(0xffffffffffffffff)
      and
      not(0xffffffffffffffff)
      and
      dup2
      mstore
      0x18
      add
```

The `not and not and` part doesn't seem to make any sense.

### PR DESCRIPTION
**Pull request by Marenz**
fixes #8670

### CODE REVIEW

**Comment by axic** [comment]
I think the test case should be added in the first commit and the second commit introduces the change *and* updates the test case, so it is clear what changed in the output.

**Comment by leonardoalt** [comment]
I'm wondering if this is the only path in our code that leads to 2 `convertType`. I couldn't find more, still wondering.

**Comment by Marenz** [comment]
What is missing here?

**Comment by axic** [comment]
@chriseth are we merging this for 0.6.12?

File: test/cmdlineTests/avoid_double_cleanup/err [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.
```
Comment (chriseth) [inline test/cmdlineTests/avoid_double_cleanup/err]: Can you try to fix this so it does not contain warnings?

Comment (chriseth) [reply test/cmdlineTests/avoid_double_cleanup/err]: Although I'm not sure if this is really the right kind of test for this issue.

File: test/cmdlineTests/avoid_double_cleanup/args [hunk 2]
```diff
@@ -0,0 +1 @@
+--asm --optimize
```
Comment (chriseth) [inline test/cmdlineTests/avoid_double_cleanup/args]: Should not use the optimizer.

File: libsolidity/codegen/CompilerUtils.h [hunk 3]
```diff
@@ -110,7 +110,7 @@ class CompilerUtils
 	/// are always padded (except for byte arrays), regardless of this parameter.
 	/// Stack pre: memory_offset value...
 	/// Stack post: (memory_offset+length)
-	void storeInMemoryDynamic(Type const& _type, bool _padToWords = true);
+	void storeInMemoryDynamic(Type const& _type, bool _padToWords = true, bool _cleanup = true);
```
Comment (chriseth) [inline libsolidity/codegen/CompilerUtils.h:114]: Please document the parameter.

File: libsolidity/codegen/CompilerUtils.h [hunk 4]
```diff
@@ -309,7 +309,7 @@ class CompilerUtils
 	void cleanHigherOrderBits(IntegerType const& _typeOnStack);
 
 	/// Prepares the given type for storing in memory by shifting it if necessary.
-	unsigned prepareMemoryStore(Type const& _type, bool _padToWords);
+	unsigned prepareMemoryStore(Type const& _type, bool _padToWords, bool _cleanup = true);
```
Comment (chriseth) [inline libsolidity/codegen/CompilerUtils.h:314]: Please document the parameter.

File: libsolidity/codegen/CompilerUtils.cpp [hunk 5]
```diff
@@ -1505,7 +1510,7 @@ unsigned CompilerUtils::prepareMemoryStore(Type const& _type, bool _padToWords)
 
 	bool leftAligned = _type.category() == Type::Category::FixedBytes;
 
-	convertType(_type, _type, true);
+	convertType(_type, _type, _cleanup);
```
Comment (chriseth) [inline libsolidity/codegen/CompilerUtils.cpp]: If `_cleanup` is false, calling this function is not needed, is it?

Comment (Marenz) [reply libsolidity/codegen/CompilerUtils.cpp]: I think you're right. I'll add an `if`

File: test/cmdlineTests/avoid_double_cleanup/err (hunk 1 above)
Comment (Marenz) [reply test/cmdlineTests/avoid_double_cleanup/err]: Do you have a suggestion for a better test?

Comment (chriseth) [reply test/cmdlineTests/avoid_double_cleanup/err]: The test should be much more focused. Maybe you could count the number of `and` opcodes or check the code generated for a specific expression only (we have other tests of both kinds).

**Comment by Marenz** [comment]
I tried adding the following test case:
```
BOOST_AUTO_TEST_CASE(avoid_double_cleanup)
{
	char const* sourceCode = R"(
		contract C {
		  function f() external {
				bytes memory amount;
				bytes24 zero_bytes24;
				bytes32 signature_root;
				bytes32 node = sha256(abi.encodePacked(
					sha256(abi.encodePacked(amount, zero_bytes24, signature_root))
				));
		  }
		}
	)";
	compileBothVersions(sourceCode, 0, "C", 50);
	// This is counting in the deployed code.
	BOOST_CHECK_EQUAL(numInstructions(m_nonOptimizedBytecode, Instruction::AND), 9);
	BOOST_CHECK_EQUAL(numInstructions(m_optimizedBytecode, Instruction::AND), 7);
}
```

The problem is the `m_nonOptimizedBytecode` check always has 9 instead of 7 and that mismatches my tests using `solc`. The resulting bytecode seems to be very different (I was only able to compare at the hex-binary output). I compared both deployed and non-deployed code from solc.
The optimized code looks more similar, matching the deployed code except for some small block which is likely some metadata..

File: test/libsolidity/SolidityOptimizer.cpp [hunk 6]
```diff
@@ -708,6 +708,20 @@ BOOST_AUTO_TEST_CASE(shift_optimizer_bug)
...
+		contract C {
+			receive() external payable {
+				abi.encodePacked(uint200(0));
+			}
+		}
+	)";
+	compileBothVersions(sourceCode, 0, "C", 50);
+	// This is counting in the deployed code.
+	BOOST_CHECK_EQUAL(numInstructions(m_nonOptimizedBytecode, Instruction::AND), 1);
+	BOOST_CHECK_EQUAL(numInstructions(m_optimizedBytecode, Instruction::AND), 0);
```
Comment (chriseth) [inline test/libsolidity/SolidityOptimizer.cpp]: This line is not needed and might be misleading into thinking that the point of the test is to check that the optimizer removes the and. Can you add a comment that we want the non-optimized code to only contain a single and?

File: test/cmdlineTests/avoid_double_cleanup/input.sol [hunk 7]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: GPLv3
```
Comment (chriseth) [inline test/cmdlineTests/avoid_double_cleanup/input.sol]: Please remove this directory.

**Comment by chriseth** [comment]
Is there a reason why this is still in draft mode?

**Comment by Marenz** [comment]
> Is there a reason why this is still in draft mode?

Nope, It was just originally you who put it in draft mode so I figured you have the criteria in mind for when to "undraft" it ;)

**Review by chriseth — APPROVED** [review]

**Pull request merged**